mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -101,7 +104,19 @@ def __init__(self, db_uri, default_artifact_root):
         self.db_uri = db_uri
         self.db_type = extract_db_type_from_uri(db_uri)
         self.artifact_root_uri = default_artifact_root
-        self.engine = mlflow.store.db.utils.create_sqlalchemy_engine_with_retry(db_uri)
+        # Quick check to see if the respective SQLAlchemy database engine has already been created.
+        if db_uri not in SqlAlchemyStore._db_uri_sql_alchemy_engine_map:
+            with SqlAlchemyStore._db_uri_sql_alchemy_engine_map_lock:
+                # Repeat check to prevent race conditions where one thread checks for an existing
+                # engine while another is creating the respective one, resulting in multiple
+                # engines being created. It isn't combined with the above check to prevent
+                # inefficiency from multiple threads waiting for the lock to check for engine
+                # existence if it has already been created.
+                if db_uri not in SqlAlchemyStore._db_uri_sql_alchemy_engine_map:
+                    SqlAlchemyStore._db_uri_sql_alchemy_engine_map[
+                        db_uri
+                    ] = mlflow.store.db.utils.create_sqlalchemy_engine_with_retry(db_uri)
+        self.engine = SqlAlchemyStore._db_uri_sql_alchemy_engine_map[db_uri]
```

Comment:
Do we want to move this assignment to the lock as well? Could we have a situation in which two threads encounter this block of code, but by the time one thread acquires the lock, creates the engine and assigns it, the other one is still acquiring the lock (because to that thread, `db_uri` is not yet in map)?

Reply:
@dbczumar and I talked about this where moving assignment inside of the lock could significantly slow down the threads since they each have to wait for the lock now. That's the purpose of the first check, which decreases the number of race conditions that could result in duplicate engines rather than eliminate. What do you think?

Reply:
Discussed offline. One extra engine probably isn't a problem :)

Reply:
@apurva-koti It could be a problem if the pool size is large. I'd rather keep the locking here, since it doesn't add much complexity.